dynamic/clientset/clientset.go

```diff
@@ -175,3 +175,32 @@ func (rc *ResourceClient) RemoveFinalizer(orig *unstructured.Unstructured, name
 		return true
 	})
 }
+
+// AtomicStatusUpdate is similar to AtomicUpdate, except that it updates status.
+func (rc *ResourceClient) AtomicStatusUpdate(resourceMap *dynamicdiscovery.ResourceMap, parentResource *dynamicdiscovery.APIResource, orig *unstructured.Unstructured, update func(obj *unstructured.Unstructured) bool) (result *unstructured.Unstructured, err error) {
+	name := orig.GetName()
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		current, err := rc.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if current.GetUID() != orig.GetUID() {
+			// The original object was deleted and replaced with a new one.
+			return apierrors.NewNotFound(rc.GroupResource(), name)
+		}
+		if changed := update(current); !changed {
+			// There's nothing to do.
+			result = current
+			return nil
+		}
+
+		if resourceMap.HasSubresource(parentResource, "status") {
```

Comment:
This should be encapsulated into `APIResource` so that you can write it as:

```go
if rc.HasSubresource("status") {
```

That way we don't require the caller to pass in extra arguments. It probably requires storing a list of subresources in the APIResource struct.